OSDOCS-3167: Test, Document and Support adding a node with a separate /var partition (#40644)
openshift/openshift-docs: modules/machine-node-custom-partition.adoc

// Module included in the following assemblies:
//
// * machine_management/
// * machine_management/
// * post_installation_configuration/node-tasks.adoc

> 

Reviewer: Consider adding the metadata that Kathryn mentioned in the last team meeting up here.

[id="machine-node-custom-partition_{context}"]

> = Optional: Adding a new {op-system} worker node with a custom `/var` partition in AWS

Reviewer: Should modules generally be marked as "Optional?" In theory, shouldn't I be able to use this module in an assembly where the procedure is mandatory?

Author: @maxwelldb I took that from [another module in the Post Install/Node Tasks assembly](https://deploy-preview-40644--osdocs.netlify.app/openshift-enterprise/latest/post_installation_configuration/node-tasks.html#rhel-removing-rhcos_post-install-node-tasks). I'm 100% sure I like having `Optional:` in the heading. But I wanted to keep the parallelism. 
If you believe it should go, I will happily remove it.

Reviewer: I believe it should go based on the spirit of mod docs, but I can't point to anything that says it is unacceptable. /shrug

> {product-title} supports partitioning devices during installation by using machine configs that are processed during the bootstrap. However, if you use `/var` partitioning, the device name must be determined at installation and cannot be changed. This means that you cannot add different instance types as nodes if they have a different device naming schema. For example, if you configured the `/var` partition with the default AWS device name for `m4.large` instances, `dev/xvdb`, you cannot directly add an AWS `m5.large` instance, as `m5.large` instances use a `/dev/nvme1n1` device by default. The device might fail to partition due to the different naming schema.   

Reviewer: My translation sense wonders if using a `/var` as an adjective for some noun might be better here. 

```suggestion
{product-title} supports partitioning devices during installation by using machine configs that are processed during the bootstrap. However, if you use `/var` [$NOUN] partitioning, the device name must be determined at installation and cannot be changed. You cannot add different instance types as nodes if they have a different device naming schema. For example, if you configured the `/var` partition with the default AWS device name for `m4.large` instances, `dev/xvdb`, you cannot directly add an AWS `m5.large` instance, as `m5.large` instances use a `/dev/nvme1n1` device by default. The device might fail to partition due to the different naming schema.   
```

The procedure in this section shows how to add a new {op-system-first} compute node with an instance that uses a different device name from what was configured at installation. You create a custom user data secret and configure a new machine set. These steps are specific to an AWS cluster. The principles apply to other cloud deployments also. However, the device naming schema is different for other deployments and should be determined on a per-case basis.

.Procedure 

> . Change to the `openshift-machine-api` namespace:

Reviewer: As I recall, ISG likes to introduce the first step for command line things explicitly, e.g. "On a command line..."

          }
        ]
      }
    },
    "version": "3.2.0"
  }
}

        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,.....=="
          }
        ]
      }
    },
    "version": "3.2.0"
  },
  "storage": {
    "disks": [
      {
        "device": "/dev/nvme1n1", <1>
        "partitions": [
          {
            "label": "var",
            "sizeMiB": 50000 <2>
            "startMiB": 0 <3>
          }
        ]
      }
    ],
    "filesystems": [
      {
        "device": "/dev/disk/by-partlabel/var", <4>
        "format": "xfs", <5>
        "path": "/var" <6>
      }
    ]
  },
  "systemd": {
    "units": [ <7>
      {
        "contents": "[Unit]\nBefore=local-fs.target\n[Mount]\nWhere=/var\nWhat=/dev/disk/by-partlabel/var\nOptions=defaults,pquota\n[Install]\nWantedBy=local-fs.target\n",
        "enabled": true,
        "name": "var.mount"
      }
    ]
  }
}
----

> //Copied from installation-disk-partitioning-upi-templates.adoc

Reviewer: Per this note, I'll skip reviewing this. :)

<4> Specifies an absolute path to the `/var` partition.
<5> Specifies the filesystem format.
<6> Specifies the mount-point of the filesystem while Ignition is running relative to where the root filesystem will be mounted. This is not necessarily the same as where it should be mounted in the real root, but it is encouraged to make it the same.
<7> Defines a systemd mount unit that mounts the `/dev/disk/by-partlabel/var` device to the `/var` partition.

.. Extract the `disableTemplating` section from the `work-user-data` secret to a text file:
+
[source,terminal]
----
$ oc get secret worker-user-data --template='{{index .data.disableTemplating | base64decode}}' | jq > disableTemplating.txt
----

.. Create the new user data secret file from the two text files. This user data secret passes the additional node partition information in the `userData.txt` file to the newly created node.
+
[source,terminal]
----
$ oc create secret generic worker-user-data-x5 --from-file=userData=userData.txt --from-file=disableTemplating=disableTemplating.txt
----

. Create a new machine set for the new node: 

> .. Create a new machine set YAML file, similar to the following, which is configured for AWS. Add the required partitions and the newly-created user data secret:

Reviewer: Similar to the following or using the following as a template?

Author: I would say _similar to the following_ as we have a note right after that suggests using an existing machine config as a template, so that you can get the correct labels and such for your cluster.

            ebs:
              encrypted: true
              iops: 0
              volumeSize: 50
              volumeType: gp2
          credentialsSecret:
            name: aws-cloud-credentials
          deviceIndex: 0
          iamInstanceProfile:
            id: auto-52-92tf4-worker-profile
          instanceType: m5.large
          kind: AWSMachineProviderConfig
          metadata:
            creationTimestamp: null
          placement:
            availabilityZone: us-east-2b
            region: us-east-2
          securityGroups:
          - filters:
            - name: tag:Name
              values:
              - auto-52-92tf4-worker-sg
          subnet:
            id: subnet-07a90e5db1
          tags:
          - name: kubernetes.io/cluster/auto-52-92tf4
            value: owned
          userDataSecret:
            name: worker-user-data-x5 <4>
----
<1> Specifies a name for the new node.
<2> Specifies an absolute path to the AWS block device, here an encrypted EBS volume.
<3> Optional. Specifies an additional EBS volume. 
<4> Specifies the user data secret file.

.. Create the machine set:
+
[source,yaml]
----
$ oc create -f <file-name>.yaml
----
+

> It can take a few moments for the machines to become available. 

Reviewer: ```suggestion
The machines may take a few moments to become available.
```

+
[source,terminal]
----
{
  "ignition": {
    "config": {
      "merge": [
        {
          "source": "https:...."
        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURFRENDQWZpZ0F3SUJBZ0lJTjRGN3RORnlWMGd3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl5TURFeE56RTVOREF4T1ZvWApEVE15TURFeE5URTVOREF4T1Zvd0pqRVNNQkFHQTFVRUN4TUpiM0JsYm5Ob2FXWjBNUkF3RGdZRFZRUURFd2R5CmIyOTBMV05oTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF3OFBRODR2SkM4WVkKL3lQK3lSeHpoMWdKa2RHcmJVRVNNzUnh5c0l6YW0rczFISGVtcS90MWN3SEh1NCtrRTNVVllCanlpdE9=="
          }
        ]
      }
    },
    "version": "3.2.0"
  }
}

        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,.....=="
          }
        ]
      }
    },
    "version": "3.2.0"
  },
  "storage": {
    "disks": [
      {
        "device": "/dev/nvme1n1", <1>
        "partitions": [
          {
            "label": "var",

>             "sizeMiB": 50000 <2>

Reviewer 2: @mburke5678 While using this I found a little typo, the trailing comma `,` is missing.

Author: Should be _**"sizeMiB": 50000,**_?

Reviewer 2: Correct! :)

Author: Fixed: https://github.com/openshift/openshift-docs/pull/43028

Reviewer 2: Cool, thanks

// Module included in the following assemblies:
//
// * machine_management/
// * machine_management/
// * post_installation_configuration/node-tasks.adoc

[id="machine-node-custom-partition_{context}"]
= Optional: Adding a new {op-system} worker node with a custom `/var` partition in AWS

> {product-title} supports partitioning devices during installation by using machine configs that are processed during the bootstrap. However, if you use `/var` partitioning, the device name must be determined at installation and cannot be changed. This means that you cannot add different instance types as nodes if they have different device naming schema. For example, if you configured the `/var` partition with the default AWS device name for `m4.large` instances, `dev/xvdb`, you cannot directly add an AWS `m5.large` instance, as it uses the `/dev/nvme1n1` device by using only a new machine set.   

Reviewer 3: ```as it uses the `/dev/nvme1n1` device by using only a new machine set.``` is a bit unclear. Maybe something like:

```as m5.large instances by default has a `/dev/nvme1n1` device instead, such that if you scale existing machinesets, it will fail to partition due to the different naming schema```

> The procedure in this section shows how to add a new {op-system-first} compute node with an instance that uses a different device name from what was configured at installation. You will create a custom user data secret and configure a new machine set. These steps are specific to an AWS cluster.

Reviewer 3: We can maybe consider adding at the end, "The principle applies to other cloud deployments as well, but the device naming schema would be different and should be determined on a per-case basis."

If we want to just document for AWS (as that was the original platform the user request came in on, what we have is fine too

.Procedure 

> . Change to the `openshift-machine-api` namespace:

Reviewer 3: Changes to `machineset` objects in the `openshift-machine-api` namespace:

Author: I was suggesting changing to the `openshift-machine-api` namespace, as the `work-user-data` secret must be in that namespace (correct?).

Reviewer 3: Ah sorry I misread, I see what you mean.

So you can switch to the machine-api namespace so you don't have to specify the namespace when interacting with objects, but the following commands, you have `-n openshift-machine-api` which essentially does the same thing if you don't switch the namespace. I'm fine with leaving this here but it is redundant I believe.

+
[source,terminal]
----

> $ oc project openshift-machine-api

Reviewer 3: I don't think this should be here

----

. Create a new secret from the the `worker-user-data` secret: 

.. Export the `userData` section of the secret to a text file: 
+
[source,terminal]
----
$ oc -n openshift-machine-api get secret worker-user-data --template='{{index .data.userData | base64decode}}' | jq > userData.txt
----

> .. Edit the text file to add the `storage`, `filesystems`, and systemd` stanzas for the partitions you want to use for the new node. You can specify any link:https://coreos.github.io/ignition/configuration-v3_3/[Ignition configuration parameters] as needed.

Reviewer 3: The ignition version is heavily dependent on OCP version. We actually ship 3.2 in 4.10

Reviewer 4: Yes, the ignition configuration parameters link should be https://coreos.github.io/ignition/configuration-v3_2/ , the one for 3.2.

I'm sorry, I gave Michael the wrong link. As you said it should be the 3.2 one.

+
[NOTE]
====

> Do not change the `ignition.config.version` stanza. 

Reviewer 3: They probably shouldn't change the `config/merge` and `security/tls/CAs` below either

    "config": {
      "merge": [
        {
          {
  "ignition": {
    "config": {
      "merge": [
        {
          "source": "https://api-int.origin-ci-int-aws.dev.rhcloud.com:22623/config/worker"
        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURFRENDQWZpZ0F3SUJBZ0lJTjRGN3RORnlWMGd3RFFZSktvWklodmNOQVFFTEJRQXdKakVTTUJBR0ExVUUKQ3hNSmIzQmxibk5vYVdaME1SQXdEZ1lEVlFRREV3ZHliMjkwTFdOaE1CNFhEVEl5TURFeE56RTVOREF4T1ZvWApEVE15TURFeE5URTVOREF4T1Zvd0pqRVNNQkFHQTFVRUN4TUpiM0JsYm5Ob2FXWjBNUkF3RGdZRFZRUURFd2R5CmIyOTBMV05oTUlJQklqQU5CZ2txaGtpRzl3MEJBUUVGQUFPQ0FROEFNSUlCQ2dLQ0FRRUF3OFBRODR2SkM4WVkKL3lQK3lSeHpoMWdKa2RHcmJVRVNNzUnh5c0l6YW0rczFISGVtcS90MWN3SEh1NCtrRTNVVllCanlpdE9=="
          }
        ]
      }
    },
    "version": "3.2.0"
  }
}

        }
      ]
    },
    "security": {
      "tls": {
        "certificateAuthorities": [
          {
            "source": "data:text/plain;charset=utf-8;base64,.....=="
          }
        ]
      }
    },
    "version": "3.2.0"
  },
  "storage": {
    "disks": [
      {

>         "device": "/dev/nvme1n1", <1>

Reviewer 3: I forget, is `nvme1n1` the boot disk?

Reviewer 4: I think the boot disk is /dev/nvme0n1

```
 oc debug node/ip-10-0-73-82.us-east-2.compute.internal -- chroot /host lsblk
Starting pod/ip-10-0-73-82us-east-2computeinternal-debug ...
To use host binaries, run `chroot /host`
NAME        MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
nvme0n1     259:0    0   120G  0 disk 
|-nvme0n1p1 259:2    0     1M  0 part 
|-nvme0n1p2 259:3    0   127M  0 part 
|-nvme0n1p3 259:4    0   384M  0 part /boot
`-nvme0n1p4 259:5    0 119.5G  0 part /sysroot
nvme1n1     259:1    0    50G  0 disk 
`-nvme1n1p1 259:6    0  48.8G  0 part /var
```

[id="machine-node-custom-partition_{context}"]
= Optional: Adding a new {op-system} worker node with a custom `/var` partition in AWS

> {product-title} supports partitioning devices during installation by using machine configs that are processed during the bootstrap. However, if you use `/var` partitioning, the device name must be determined at installation and cannot be changed. This means that you cannot add different instance types as nodes if they have different device naming schema. For example, if you configured the `/var` partition with the default AWS device name for `m4.large` instances, `dev/xvdb`, you cannot directly add an AWS `m5.large` instance, as m5.large instances use a `/dev/nvme1n1` deviceby default. If you scale an existing machine set, it will fail to partition due to the different naming schema.   

Author: Thank you!
